Allows migration of DVR routers via evacuate

aplanas opened the pull request.

cloudnull committed:
Updated the neutron l3HA tool to use v3

The change adds Keystone v3 support to neutron l3 ha tool. This
change is needed to ensure that all l3-ha fail overs are
accomplished correctly.

The tool has been updated to skip / support routers that have
already bene setup using the built in neutron L3HA.

This is something that will need to be backported to Kilo

Change-Id: I94f21ff3fa511bc9da6b57cdf847719620818ece
Closes-Bug: #1499708
Signed-off-by: Kevin Carter <kevin.carter@rackspace.com>

https://bugs.launchpad.net/openstack-ansible/+bug/1499708
https://review.openstack.org/#/c/229053/

aplanas committed:
neutron-ha-tool: order imports

Ordering imports makes more easy backport patches from the
ansible[1] playbook repository.

[1] git://git.openstack.org/openstack/openstack-ansible.git

aplanas committed:
neutron-ha-tool: add spaces inside expressions.

aplanas committed:
neutron-ha-tool: filter agents based on agent_mode value (bsc#974495)

`target_agent_list` is filtering the list of agents excluding
the one that we want to migrate from.  We need to expand this
exclusion to the agents that contains a different configuration
agent_mode parameter.

For example, if we want to migrate from one `dvr_snat` agent to
another one, we need to exclude the `dvr`s one.

https://bugzilla.suse.com/show_bug.cgi?id=974495

aplanas commented:
@aspiers @vuntz r?

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -43,14 +45,19 @@
-            # We don't have a logger set up yet
-            sys.stderr.write("Your environment is missing '%s'\n" % key)
-            sys.exit(1)
+            raise SystemExit("Your environment is missing '%s'")
```
Missing `% key`?

aspiers replied:
OK nice, so this is a rare valid exception to the general rule that PRs shouldn't deliberately introduce breakage and then immediately fix it.  In this case I agree it is worth cherry-picking the change from the ansible repo unchanged (except for the additions to the commit message I have suggested) and then fixing issues with the cherry-picked commit in separate commits.

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -43,14 +45,19 @@
-    for key in ['OS_TENANT_NAME', 'OS_USERNAME',
-                'OS_AUTH_URL', 'OS_REGION_NAME']:
+    for key in ['OS_USERNAME', 'OS_PASSWORD', 'OS_AUTH_URL',
+                'OS_ENDPOINT_TYPE']:
```
Is it intended to lose `OS_ENDPOINT_TYPE`? Also, we need to check that `OS_PROJECT_NAME` or `OS_TENANT_NAME` is there.

aplanas replied:
Ah right. This comes from the backport. I tested it manually via command line, with the source .openrc and with the password file. Will recheck via RA.

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -160,15 +167,38 @@ def run(args):
+    kclient_kwargs['password'] = os.environ['OS_PASSWORD']
+    kclient_kwargs['insecure'] = args.insecure
+    kclient_kwargs['ca_cert'] = ca
+    kclient_kwargs['auth_url'] = os.environ['OS_AUTH_URL']
```
We lost the region name here...

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -160,15 +167,38 @@ def run(args):
+        '2.0',
+        endpoint_url=keystone.service_catalog.url_for(
+            service_type='network',
+            endpoint_type=os.environ['OS_ENDPOINT_TYPE']
```
I'm nearly sure that our crowbar cookbook doesn't set this environment variable in the pacemaker resource (old code was using `internalURL` hard-coded, it seems).

aspiers replied:
https://github.com/crowbar/crowbar-openstack/blob/master/chef/cookbooks/neutron/recipes/network_agents_ha.rb#L184

aplanas commented:
@vuntz addressed the issues, and tested in SOC7 restarting the RA

In SOC6 there is a problem in the neutron code that makes the manual and the automatic migration unreliable. I need to find the issue and backport it into the package.

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -21,36 +21,47 @@
     # ensure environment has necessary items to authenticate
-    for key in ['OS_TENANT_NAME', 'OS_USERNAME',
-                'OS_AUTH_URL', 'OS_REGION_NAME']:
+    for key in ['OS_USERNAME', 'OS_PASSWORD', 'OS_AUTH_URL',
```
`OS_PASSWORD` is not required as it can be passed through a file.

aplanas replied:
I know, but is set in the RA file. Is empty but is always there.

vuntz reviewed (changes requested):
One additional fix required -- I hadn't noticed that earlier.

Everything else looks good!

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -31,8 +31,10 @@
 import sys
```
[Since github does not let me review commit messages (another reason #githubsucks), I'm reviewing it here.]

Please can you update this commit message to explain where it was cherry-picked from, and also add these links:
- https://bugs.launchpad.net/openstack-ansible/+bug/1499708
- https://review.openstack.org/#/c/229053/

aspiers replied:
These are the links which I was asking to be added to this backport commit ("Updated the neutron l3HA tool to use v3")

aplanas replied:
Both links added. Done.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -737,10 +737,21 @@ def target_agent_list(agent_list, agent_type, exclude_agent_host):
     :param exclude_agent_host: hostname of agent to exclude from the list
 
     """
+    agent_info = [
```
Would be a bit clearer if this was called `excluded_agent_info`, or `source_agent_info`.

aplanas replied:
IMHO is more clear `agent_info` here: shorter name, direct and refers to what contains and not from who. Also the variable is local, and the usage is close.

aspiers replied:
Shorter does not imply clearer; in fact IMHO it's the exact opposite in this case: this function is dealing with multiple agents, so it's ambiguous which agent `agent_info` refers to.  The only way to understand which agent it refers to is to read the code which assigns to it.  (I'm not sure exactly what you mean by "not from who"?)

In contrast, if it was called `excluded_agent_info` then it would be self-evident, because it would obviously be referring to the docstring which says "Return a list of agents that are alive, **excluding** the one we want to migrate from".

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -737,10 +737,21 @@ def target_agent_list(agent_list, agent_type, exclude_agent_host):
     :param exclude_agent_host: hostname of agent to exclude from the list
```
Please can you get in the habit of including full bugzilla URLs in the commit message, as per [team policy](https://etherpad.nue.suse.com/p/cloud-pull-requests)?  In this case it would be https://bugzilla.suse.com/show_bug.cgi?id=974495

aplanas replied:
Why the comment is in the line of code? Where do you want that I add the full link of the bug? The bug is stated in the commit message, in such a way that can be listed in the changelog.

I am not aware of this team policy etherpad, and I am not sure that OBS / IBS can track the full URL in the commit description. I would prefer to add this bsc#XXXXX signature for convenience in the commit message.

aspiers replied:
> Why the comment is in the line of code?

Because [github doesn't allow reviewing of commit messages](https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/6#discussion_r88211081).

> Where do you want that I add the full link of the bug?

In the commit message, like I said.

> The bug is stated in the commit message, in such a way that can be listed in the changelog.

Yes, but the _full URL_ is missing.

> I am not aware of this team policy etherpad

You are now, since I told you ;-)

> and I am not sure that OBS / IBS can track the full URL in the commit description.

It doesn't need to; it can track the `bsc#XXXXXX` number.

> I would prefer to add this bsc#XXXXX signature for convenience in the commit message.

I didn't ask you to drop the `bsc#XXXXXXX`.  In fact the etherpad requires that, and IMHO it's exactly right that you have already put the `bsc#XXXXXXX` in the first line of the commit message.  I was asking for the full URL to be included in the commit message body, in _addition_ to that.  The justification is given in the etherpad.

aplanas replied:
Added, but I disagree with the etherpad and the justification is weak. For one side nearly all projects refers to bugs using the local prefix associated with namespace of the project, or even with a single number. For the other side our commits from other project do not follow this rule.

> > I am not aware of this team policy etherpad
> 
> You are now, since I told you ;-)

No. Not really. When this policy was discussed? Who participate on it? When has been announced and communicated? I am not able to find this in the mailing list.

aspiers replied:
> Added, but I disagree with the etherpad and the justification is weak.

Please can you explain exactly why you think it is weak?  The two reasons given in the etherpad are:
1. this makes the reference clear to outsiders who don't know what bnc / bsc is
2. it makes it possible to open the bug with a single click when viewing git history

I really struggle to imagine how you could disagree with either of these.  For the first one, remember that this is a public repository, and the code relates to a fork of some upstream code which is still in use elsewhere.  So any comments about `bsc#XXXXXX` being sufficient in internal or openSUSE-specific tools are irrelevant.  And the second one is a simple fact.

>  For one side nearly all projects refers to bugs using the local prefix associated with namespace of the project, or even with a single number. For the other side our commits from other project do not follow this rule.

Again, I am not suggesting to drop the `bsc#XXXXX` format, so how is this relevant?  I am only saying that it is worth adding the full URL in _addition_, because this brings extra clarity and convenience for anyone who reads the PR and/or the git commits.  It's not _needed_, but it is a very nice courtesy which takes very little effort and makes other people's lives easier, so why not do it?  It follows the ethos of [this great article on Empathy and Pull Requests](https://slack.engineering/on-empathy-pull-requests-979e4257d158), which I recommend reading if you didn't already.

> No. Not really.

You _are_ aware of it, even if you don't agree with it :-)

> When this policy was discussed? Who participate on it? When has been announced and communicated? I am not able to find this in the mailing list.

It was discussed a long time ago across several team workshops in Nuremberg, possibly before you joined the team, but I can't remember exactly when.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -419,7 +425,7 @@ def l3_agent_migrate(qclient, noop=False, now=False):
              total_migrations)
     if total_errors > 0:
-        LOG.error("%d errors encountered during migration")
+        LOG.error("%d errors encountered during migration", total_errors)
```
These were from #5 so I'm not sure why they are in this unrelated commit.  However I have now merged #5 so they will vanish on rebase.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -21,16 +21,16 @@
 #   2 - router migration required (with --l3-agent-check)
```
Would be helpful to include the URL of the ansible playbook repo in the commit message.

aplanas replied:
Done!

aspiers replied:
Thanks, but you added the URL to a different commit.  I was suggesting to add it to this commit ("neutron-ha-tool: order imports") because the commit message is:

> Ordering imports makes more easy backport patches from the
> ansible playbook repository.

so anyone reading it will immediately wonder which repo that is.

aplanas replied:
Makes sense. Done.

aspiers replied:
Great, thanks!

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -54,10 +54,13 @@
     # ensure environment has necessary items to authenticate
-    for key in ['OS_USERNAME', 'OS_PASSWORD', 'OS_AUTH_URL',
-                'OS_ENDPOINT_TYPE']:
+    for key in ['OS_USERNAME', 'OS_AUTH_URL', 'OS_REGION_NAME']:
```
Please make the commit message explain why we're dropping `OS_ENDPOINT_TYPE`.

aplanas replied:
I will.

aplanas replied:
Done!

aspiers replied:
Thanks for the commit message update! Ahh, I misunderstood - we're not actually dropping `OS_ENDPOINT_TYPE`, but merely making it optional.  So the commit message is still slightly misleading.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -171,22 +174,25 @@ def run(args):
         auth_version = os.getenv('OS_IDENTITY_API_VERSION', None)
         if not auth_version:
-            auth_version = 3
+            auth_version = '2.0'
```
I don't understand why we've cherry-picked a commit to add v3 support, and then immediately regress to v2.0?  If this is really correct then the commit message needs to explain why.

aplanas replied:
We add v3 from the commit, but we fallback to v2 if the version is not specified.

aspiers replied:
Got it now - thanks!

aspiers reviewed (changes requested):
Thanks, the code changes basically look good but I think the commit messages could still be improved a bit as it's not 100% clear yet.

aplanas commented:
@aspiers Updated the comments in the commits

vuntz commented:
@aplanas https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/5 got merged, so there's now a conflict. Can you rebase?

aplanas commented:
@vuntz  rebased, thanks!

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -419,7 +425,7 @@ def l3_agent_migrate(qclient, noop=False, now=False):
              total_migrations)
     if total_errors > 0:
-        LOG.error("%d errors encountered during migration" % total_errors)
+        LOG.error("%d errors encountered during migration", total_errors)
```
These changes are not related to this commit.  If there is a good reason for them then I'm fine with that, but please put them in a separate commit with a message explaining the change.

aplanas replied:
Moved to a different commit. Done

aspiers reviewed (changes requested):
Almost there now, thanks!  Still a couple of minor nitpicks around commit hygiene, but the main change I would like to see is adding those two URLs I suggested into the backport of the "Updated the neutron l3HA tool to use v3" commit, so that it's easier to see where it came from.

aplanas committed:
neutron-ha-tool: adjust OS envars as used with current RA

* Do not require OS_PASSWORD. Resource agent set this
  variable, but we want to be sure to read it from the
  file.

* Do not require OS_ENDPOINT_TYPE. Not used in the
  resource agent.
  Default it to `internalURL` are requested in the RA.

* Default to v2.0 of keystone if OS_IDENTITY_API_VERSION
  is not set.

aplanas committed:
neutron-ha-tool: change log interpolation var

As documented[1], variables for interpolation in the log message
are expected in the `args` or `kwargs` parameter.

[1] https://docs.python.org/2/library/logging.html#logging.debug

aspiers reviewed (approved):
Excellent!  Thanks very much for your patience with my pedantry :-)

aspiers commented:
This just reminded me that we have no CI coverage for this code :-/ I guess we'll have to just trust manual testing for now...

aplanas commented:
@aspiers I am very grateful for the work that you put in the reviews.

I tested it in SOC6 and SOC7 but yes, is not a trusty source. Maybe QA can fix this. @vmoravec @gosipyan @kbaikov?
